Feature/06 app setup

KlugeCh opened the pull request:
@taitruong @RouvenGonzalez 
Please review the basic setup for the website app.
Basic App setup needs to be merged to develop branch and later in main branch to continue with deployment.

basic setup containing:

- installed express.js
- installed and configured ESLint
- installed and configured Prettier
- installed and configured eslint-config-prettier
- configured .gitignore
- installed and configured Jest for unit tests
- installed and configured Playwright for e2e tests
- installed eslint-plugin-jest-playwright
- installed jest-playwright-preset
- added README with description and instructions
- installed express-generator

For testing / running the app local, please check the README.md

KlugeCh committed:
#6 installed and configured:
install Node.js
install express.js
install and configure ESLint
install and configure Prettier
configure gitignore
install and configure Jest for unit tests
install and configure Playwright for e2e tests
installed express-generator

KlugeCh committed:
#6 added README with short description

KlugeCh committed:
#6 added script in package.json for jest test

KlugeCh committed:
#6 created jest.e2e.config.js for jest-playwright-preset and RegEx for e2e tests

KlugeCh committed:
#6 fixed linter errors in
app.js
routes/index.js
routes/users.js

KlugeCh committed:
#6 created app.yaml for deploying to google cloud

KlugeCh committed:
#6 improved README.md with description to run app local

KlugeCh committed:
#6 corrected wrong header and title in routes/index.js

KlugeCh committed:
#6 added further information to README.md

RouvenGonzalez reviewed (approved):
Sieht gut aus so als Basis :+1: 
Eventuell brauchen wir zu nem späteren Zeitpunkt eine andere Ordnerstrucktur als das was der Generator erstellt hat, aber darum können wir uns kümmern wenn es soweit ist.

Commit:
review documentation

taitruong commented on .eslintrc.json:
```diff
@@ -0,0 +1,19 @@
+    "ecmaVersion": 12
+  },
+  "rules": {
+    "indent": ["error", "tab"],
```
Please check and test: rule should be: 2 spaces (and no tab)

KlugeCh replied:
According to https://eslint.org/docs/rules/indent#rule-details, ESLint shows an 'error' if indentation is not equal to 'tab',
'tab' default style is 4 spaces.
Fixed, changed to "indent": ["error", 2], for 2 space indentation.

KlugeCh replied:
Corrected rule in .prettierrc, indentation is now 2 spaces.

taitruong commented on .vscode/settings.json:
```diff
@@ -0,0 +1,15 @@
+      "source.fixAll.eslint": true
+    },
+    "eslint.alwaysShowStatus": true,
+    "files.autoSave": "onFocusChange"
```
What does that mean?

KlugeCh replied:
- Tested:
Current file is auto saved by changing to another file.
Can be deleted if this behavior is not wanted.
- Description:
https://code.visualstudio.com/docs/editor/codebasics#_save-auto-save

taitruong commented on app.js:
```diff
@@ -0,0 +1,44 @@
+var createError = require('http-errors');
```
TODO: next story: convert to typescript, no JS

KlugeCh replied:
Created Issue for converting to typescript.

taitruong commented on package.json:
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "website",
+  "version": "1.0.0",
```
semantic versioning: "1.0.0" means this is an official release. Use another version.

KlugeCh replied:
Changed to 0.0.1-alpha.1 for marking as pre-release, according to semver:
- https://softwareengineering.stackexchange.com/questions/275358/version-number-for-a-pre-release-software-that-will-be-the-second-major-release

taitruong commented on package.json:
```diff
@@ -0,0 +1,37 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/software-developer-org/website.git"
```
What is that? This URL is unknown

KlugeCh replied:
URL corrected and tested, know working

taitruong commented on package.json:
```diff
@@ -0,0 +1,37 @@
+  "dependencies": {
+    "cookie-parser": "~1.4.4",
+    "debug": "~2.6.9",
+    "eslint": "^7.19.0",
```
Some packages should be defined as devDependencies

KlugeCh replied:
Splitted in dependencies and devDependencies.
 
"dependencies" are set by default through express-generator 
-   "cookie-parser"
 -   "debug"
 -   "express"
 -   "http-errors"
 -   "morgan"
 -   "pug"
  
"devDependencies":
 -   "eslint"
 -   "eslint-config-prettier"
 -   "eslint-plugin-jest-playwright"
 -   "jest"
 -   "jest-playwright-preset"
 -   "nodemon"
 -   "playwright"
 -   "prettier"

taitruong commented on public/stylesheets/style.css:
```diff
@@ -0,0 +1,8 @@
+body {
```
CSS and not SCSS?

KlugeCh replied:
CSS file was auto generated by express-generator.
Changing to SCSS is part of https://github.com/software-developer-org/website/issues/8

taitruong commented on routes/index.js:
```diff
@@ -0,0 +1,10 @@
+var express = require('express');
```
Next story: typescript

KlugeCh replied:
Created Issue for converting to typescript.

taitruong commented on README.md:
````diff
@@ -0,0 +1,25 @@
+# ... or start in debug mode
+DEBUG=website:* npm start
+# now open localhost:3000 in your browser
+```
````
A chapter for folder structure would be good. Also maybe some resources to express. Also maybe a simple explanation about the key files (e.g. templates).

RouvenGonzalez commented on .eslintrc.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "env": {
```
test kommentar

taitruong replied:
Ich kann es sehen

KlugeCh committed:
#6 changed .eslintrc.json to 2 space indentation

KlugeCh committed:
#6 package.json:
corrected and tested repository url
moved packages from dependencies to devDependencies:
eslint
eslint-config-prettier
eslint-plugin-jest-playwright
jest
jest-playwright-preset
nodemon
playwright
prettier

KlugeCh committed:
#6 merged improved README.md

KlugeCh committed:
#6 corrected indentation rule in .prettierrc to be 2 spaces now

KlugeCh committed:
#6 corrected versioning with pre-release tag to 1.0.0.alpha.1

KlugeCh committed:
#6 added express-generator folder structure to README.md

KlugeCh committed:
#6 added TODOs for kex files ans deploying to README.md

KlugeCh committed:
#6 changed versioning to 0.0.1-alpha.1

taitruong commented on package.json:
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "website",
+  "version": "0.0.1-alpha.1",
```
@KlugeCh, ist das die korrekte Syntax bei Sematic Versioning? 0.0.0-alpha.1 habe ich noch nie gesehen.

KlugeCh replied:
@taitruong 0.0.1-alpha ist korrekt
